src/gluonts/itertools.py

```diff
@@ -129,18 +149,79 @@ def __len__(self) -> int:
         return len(self.iterable)
 
 
-class IterableSlice(Iterable):
+# can't make this a dataclass because of pytorch-lightning assumptions
+class IterableSlice:
     """
     An iterable version of `itertools.islice`, i.e. one that can be iterated
-    over multiple times.
+    over multiple times:
+
+        >>> isl = IterableSlice([1, 2, 3, 4, 5], 3)
+        >>> list(isl)
+        [1, 2, 3]
+        >>> list(isl)
+        [4, 5]
+        >>> list(isl)
+        []
+
+    This needs to be a class to support re-entry iteration.
     """
 
-    def __init__(self, iterable: Iterable, length: Optional[int]) -> None:
+    def __init__(self, iterable, length):
         self.iterable = iterable
         self.length = length
+        self.next()
+
+    def next(self):
+        self.current = list(itertools.islice(self.iterable, self.length))
 
     def __iter__(self):
-        return itertools.islice(self.iterable, self.length)
+        yield from self.current
+        self.next()
 
     def __len__(self) -> int:
-        return len(self.iterable)
+        return len(self.current)
+
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+def rows_to_columns(
```

Comment:
These two are very simple, but we should probably have a few test cases

Reply:
They have a simple test case in case of the doctest!

Reply:
Do those run? I cannot remember honestly :-)